### DISCUSSION TITLE
ReplayGain with Opus files

### DISCUSSION THREAD

**Discussion by gahabana** [MoleMan1024/audiowagon#98] [General]
Hi, 
    I am using Opus format for my car music library. 
    ReplayGain is great for car. However, I am not sure how AudioWagon treats it with Opus files. Here is what I have read and need your advice how AudioWagon is treating it (or not):
1. Each Opus file has 'Playback Gain' in it's header. Could be 0db or value that converter/tagger puts there. All players must respect that tag and when decoding have to respect it.
2. Some converters put in that field 0, some put replaygain_track_gain and some replaygain_album_gain. E.g. ffmpeg will put 0, opusenc will put equivalent of replay_track_gain (i believe) etc
3. replaygain tags as in mp3, flac files etc are NOT allowed to be in Opus files. However R128_TRACK_GAIN and R128_ALBUM_GAIN are allowed/encouraged and player should take that into consideration.
4. **How does (if at all) AudioWagon treats Opus replaygain information ?**

**Comment by MoleMan1024** [comment]
@gahabana : you can find the implementation [here](https://github.com/MoleMan1024/audiowagon/blob/0f4ba916ad2411696b7ca54cc23b6d8460461008/automotive/src/main/java/de/moleman1024/audiowagon/player/AudioSession.kt#L107). Currently AudioWagon looks for a plain string "replaygain_track_gain" followed by the decibels of the plus/minus gain in the first and last 1024 bytes of the audio file. That seems to work for .mp3 and .flac files.

I have no experience with opus files. Maybe you could share a small file or find a link to a file that has such replaygain data, then I can check if it can be extracted.

**Reply by gahabana** [reply]
hi, will post the file in a second. from 
If a player chooses to make use of the TRACK_GAIN, it MUST be applied in addition to the OpusHead output gain. from Opus specs this may help [https://wiki.xiph.org/OggOpus] :
`If a player chooses to make use of the TRACK_GAIN, it MUST be applied in addition to the OpusHead output gain. `

Therefore no Opus file should contain fields REPLAYGAIN_TRACK_GAIN nor REPLAYGAIN_ALBUM_GAIN (and using 2 different converters/taggers I confirm that is the case) ... reference encoder (opusenc) takes those fields from FLAC file and changes them to needed (R128_TRACK_GAIN and/or R128_ALBUM_GAIN).

So the tag names are (little) different in Opus files compared to FLAC and MP3. The other difference is that with Opus reference loudness level is 'lower' by 5dB than in FLAC/MP3 etc ... so if replaygain_track_gain is -5dB in flac file in opus it will be -10dB.

In my experience - Opus files have 'gain' in their header and then if one wants to use 'replaygain_track_gain' it has to look for fields R128_TRACK_GAIN and applyformula -> ResultingGain = OpusHeaderGain + R128_TRACK_GAIN

Am struggling with 'shortening' files that would have both album and track gains but here is the output of mediainfo and metaflac for source file and mediainfo and opusinfo for resulting Opus file:

in this specific example - REPLAYGAIN_TRACK_GAIN is +1.74dB and player will actually iNCREASE volume by those 1.74dB.
In Opus file Header Gain is -3.17969 and R128_TRACK_GAIN is '-20' (as that is stored as Q7.8 format one basically divides by 256 to get floating point value = -20/256 = -0.078125 .... then substracting that from Header Gain = -3.17969 + - 0.078125 = -3.257815 that is exactly 5dB difference to original FLAC file (5 - 3.257815 = 1.742185).

Hope this helps. Can provide source/target files via DM if needed.

**Source FLAC:**

**mediainfo**

```
mediainfo 07.Drive\ -\ Driven.flac
General
Complete name                            : 07.Drive - Driven.flac
Format                                   : FLAC
Format/Info                              : Free Lossless Audio Codec
File size                                : 23.9 MiB
Duration                                 : 4 min 15 s
Overall bit rate mode                    : Variable
Overall bit rate                         : 786 kb/s
Album replay gain                        : 1.82 dB
Album replay gain peak                   : 0.690216
Album                                    : Baby
Track name                               : Drive / Driven
Track name/Position                      : 07
Track name/Total                         : 10
Performer                                : Yello
Genre                                    : Electronic
Recorded date                            : 1991
Cover                                    : Yes
Cover type                               : Cover (front)
Cover MIME                               : image/jpeg
Comment                                  : ExactAudioCopy v0.99pb5
ALBUM DYNAMIC RANGE                      : 13
DYNAMIC RANGE                            : 14
ZRATING                                  : 5

Audio
Format                                   : FLAC
Format/Info                              : Free Lossless Audio Codec
Duration                                 : 4 min 15 s
Bit rate mode                            : Variable
Bit rate                                 : 781 kb/s
Channel(s)                               : 2 channels
Channel layout                           : L R
Sampling rate                            : 44.1 kHz
Bit depth                                : 16 bits
Compression mode                         : Lossless
Replay gain                              : 1.74 dB
Replay gain peak                         : 0.625122
Stream size                              : 23.8 MiB (99%)
Writing library                          : libFLAC 1.3.1 (UTC 2014-11-25)
```

**metaflac**
```
$ metaflac --list --block-type=VORBIS_COMMENT 07.Drive\ -\ Driven.flac

METADATA block #2
  type: 4 (VORBIS_COMMENT)
  is last: false
  length: 392
  vendor string: reference libFLAC 1.3.1 20141125
  comments: 15
    comment[0]: ALBUM=Baby
    comment[1]: ALBUM DYNAMIC RANGE=13
    comment[2]: ARTIST=Yello
    comment[3]: COMMENT=ExactAudioCopy v0.99pb5
    comment[4]: DATE=1991
    comment[5]: DYNAMIC RANGE=14
    comment[6]: GENRE=Electronic
    comment[7]: TITLE=Drive / Driven
    comment[8]: TRACKTOTAL=10
    comment[9]: TRACKNUMBER=07
    comment[10]: ZRATING=5
    comment[11]: replaygain_album_gain=+1.82 dB
    comment[12]: replaygain_album_peak=0.690216
    comment[13]: replaygain_track_gain=+1.74 dB
    comment[14]: replaygain_track_peak=0.625122
```

**Target Opus file:**
```
$ mediainfo ~/out.opus
General
Complete name                            : out.opus
Format                                   : Ogg
File size                                : 3.36 MiB
Duration                                 : 4 min 15 s
Overall bit rate                         : 111 kb/s
Album                                    : Baby
Track name                               : Drive / Driven
Track name/Position                      : 07
Track name/Total                         : 10
Performer                                : Yello
Genre                                    : Electronic
Recorded date                            : 1991
Writing application                      : opusenc from opus-tools 0.1.10
Cover                                    : Yes
Comment                                  : ExactAudioCopy v0.99pb5
ALBUM DYNAMIC RANGE                      : 13
DYNAMIC RANGE                            : 14
ZRATING                                  : 5
R128_TRACK_GAIN                          : -20

Audio
ID                                       : 1233262329 (0x498216F9)
Format                                   : Opus
Duration                                 : 4 min 15 s
Channel(s)                               : 2 channels
Channel layout                           : L R
Sampling rate                            : 44.1 kHz
Compression mode                         : Lossy
Writing library                          : libopus 1.3.1
```

**Opusinfo**

```
$ opusinfo ~/out.opus
Processing file "~/out.opus"...

New logical stream (#1, serial: 498216f9): type opus
Encoded with libopus 1.3.1
User comments section follows...
	ENCODER=opusenc from opus-tools 0.1.10
	METADATA_BLOCK_PICTURE=3|image/jpeg||600x592x24|<138886 bytes of image data>
	ALBUM=Baby
	ALBUM DYNAMIC RANGE=13
	ARTIST=Yello
	COMMENT=ExactAudioCopy v0.99pb5
	DATE=1991
	DYNAMIC RANGE=14
	GENRE=Electronic
	TITLE=Drive / Driven
	TRACKTOTAL=10
	TRACKNUMBER=07
	ZRATING=5
	R128_TRACK_GAIN=-20
Opus stream 1:
	Pre-skip: 312
	Playback gain: -3.17969 dB
	Channels: 2
	Original sample rate: 44100Hz
	Packet duration:   20.0ms (max),   20.0ms (avg),   20.0ms (min)
	Page duration:   1000.0ms (max),  996.6ms (avg),  140.0ms (min)
	Total data length: 3527532 bytes (overhead: 5.98%)
	Playback length: 4m:15.120s
	Average bitrate: 110.6 kb/s, w/o overhead: 104 kb/s
Logical stream 1 ended
```

**Reply by gahabana** [reply]
In addition reference encoder ('Opusenc') if source file has NO replaygain information does nothing - it keeps output gain at 0 and does not write any tags in the output file. 
However if e.g. source file (lossless FLAC) has replaygain tags it uses Album Gain to set Opus Header Gain (so that all songs within an album have same perceived volume as other albums) ... and it only uses R128_TRACK_GAIN should one want to have TRACK_GAIN rather then ALBUM_GAIN applied on the output. 

Based on my understanding how AudioWagon works - it really means in case ReplayGain is enabled in preferences, it should look for R128_TRACK_GAIN rather then REPLAYGAIN_TRACK_GAIN in tags and thats it. Not sure which library in AAOS does decoding of Opus but my bet is that it conforms to standards and applies Opus Header Gain if it is not 0 ... so should be small and worthwhile change in AudioWagon for those of us who prefer ( in the car) TRACK_GAIN as opposed to ALBUM_GAIN.

The only programming 'trick' is that R128_TRACK_GAIN is not stored as text in dB but rather as text representing signed integer and to get value in dB one needs to divide it by 256 and resulting float . E.g. if tag is reading '892' track_gain is 892/256 is positive 3.48dB and if it is -892 then it is negative -3.48dB.

Hope this helps a bit ?

thank you !

**Reply by gahabana** [reply]
[Archive.zip]
(https://github.com/MoleMan1024/audiowagon/files/10247066/Archive.zip)
here is ZIP file with 2 'songs' properly tagged. FLAC file has replaygain tags and Opus file is created by 'opusenc' ... to read 'Opus Header Gain' you need foobar2000 on windows or 'opusinfo' command line tool availabe for all platforms. As for R128_TRACK_GAIN in Opus file it is readable by any/most tools.

**Comment by MoleMan1024** [comment]
@gahabana : thanks for all the info, I will have a look.
